core/src/main/scala/dx/core/languages/cwl/CwlBlock.scala

```diff
@@ -349,46 +354,47 @@ object CwlBlock {
                    inheritedRequirements: Vector[Requirement] = Vector.empty,
                    inheritedHints: Vector[Hint] = Vector.empty): Vector[CwlBlock] = {
     val wfInputs: Map[DxName, WorkflowInputParameter] = wf.inputs.collect {
-      case i if i.hasName =>
-        CwlDxName.fromSourceName(i.name) -> i
+      case i if i.hasName => CwlDxName.fromSourceName(i.name) -> i
     }.toMap
 
     // sort steps in dependency order
     def orderSteps(
+        wfId: Identifier,
         steps: Vector[WorkflowStep],
         deps: SeqMap[String, WorkflowStep] = SeqMap.empty
     ): SeqMap[String, WorkflowStep] = {
       val (satisfied, unsatisfied) =
         steps.partition { step =>
           step.inputs.forall { inp =>
             inp.sources.forall {
-              case id if id.parent.isDefined => deps.contains(id.parent.get)
+              case id if id.parent.contains(wfId.frag)   => true
+              case id if id.parent.exists(deps.contains) => true
               case id =>
-                wfInputs.contains(CwlDxName.fromSourceName(id.name.get))
+                wfInputs.contains(CwlDxName.fromSourceName(id.name))
             }
           }
         }
       if (satisfied.isEmpty) {
         throw new Exception("could not satisfy any dependencies")
       }
-      val newDeps = deps ++ satisfied.map(step => step.name -> step).to(SeqMap)
+      val newDeps = deps ++ satisfied.map(step => step.frag -> step).to(SeqMap)
       if (unsatisfied.isEmpty) {
         newDeps
       } else {
-        newDeps ++ orderSteps(unsatisfied, newDeps)
+        newDeps ++ orderSteps(wfId, unsatisfied, newDeps)
```

Comment:
@jdidion when unsatisfied is not empty, we call the orderSteps again, but those steps will still be recognized as unsatisfied this time, right?

Reply:
`deps = newDeps` will have added contents since the last call, so there is the chance that in each successive call, some more items will get partitioned into `satisfied` via 

```
case id if id.parent.exists(deps.contains) => true
```

Reply:
Oh just realized it is not looping within one step. Thanks Riva!

Reply:
no, because the dependencies have been updated